Add RSKIP for Segwit compatibility (#143)
rsksmart/RSKIPs: IPs/RSKIP143.md

# Segwit Compatibility

|RSKIP          |143          |
| :------------ |:-------------|
|**Title**      |Segwit Compatibility|
|**Created**    |22-OCT-19 |
|**Author**     |PGP |
|**Purpose**    |Sca,USa |
|**Layer**      |Core |
|**Complexity** |2 |
|**Status**     |Draft|

## Abstract

> This document describes the procedure allows the 2wp proccess to accept segwit transaction sent using a P2SH-PWPKH wallet. 

Reviewer: This document describes the procedure ~~allows~~ **to allow** the 2wp pro~~c~~cess to accept segwit transaction**s** sent using a P2SH-PWPKH wallet.

> This document describes the procedure to allow the 2wp proccess to accept segwit transactions sent using a P2SH-PWPKH wallet. 

Reviewer 2: ```suggestion
This document describes the procedure to change the 2wp proccess to accept segwit transactions sent using a P2SH-PWPKH wallet. 
```

## Motivation

> The 2wp proccess only allows transaction who has been sent using a P2PK address. This RSKIP describes how Bridge could accept P2SH-PWPKH[1] since this is one of the most popular wallet nowadays and the segwit transactions needs less space in a block. 

Reviewer: The 2wp ~~proccess~~ process **currently accepts only** transaction**(s)** ~~who~~ which have been sent using a P2PK **(P2PK or P2PKH?)** address. This RSKIP describes how **the** Bridge could accept P2SH-PWPKH **(P2PKH?)**[1] since this is one of the most popular wallet**(s)** nowadays and ~~the~~ segwit transactions ~~needs~~ **use** less space in a block.

> The 2wp proccess only allows transactions has been sent using a P2PKH address. This RSKIP describes how Bridge could accept P2SH-PWPKH[1] since this is one of the most popular wallet nowadays and the segwit transactions uses less space in a block. 

Reviewer 2: ```suggestion
The 2wp proccess only accepts transactions that were sent using a P2PKH address. This RSKIP describes how Bridge could accept P2SH-PWPKH[1] addreses since this is one of the most popular wallet nowadays and the segwit transactions use less space in a block. 
```

> The 2wp proccess only accepts transactions that were sent using a P2PKH address. This RSKIP describes how Bridge could accept P2SH-PWPKH[1] addreses since this is one of the most popular wallet nowadays and the segwit transactions use less space in a block. 

Reviewer: Typo: addreses -> **addresses**

## Specification

During the execution of the process, it parse each new transaction sent to obtain public key from it. If this identification cannot be made, Bridge considers the transaction as invalid.

> The changes related to this RKSIP refer to changing the Bridge code of RegisterBtcTransactions function. This function is responsible to parse the transactions and obtaining the public key value of the data sent to it. 

Reviewer: This function is responsible ~~to~~ of parsing the transactions and obtaining the public key value ~~of~~ from the data sent ~~to it~~.

> The changes related to this RKSIP refer to changing the Bridge code of RegisterBtcTransactions function. This function is responsible for parsing Bridge transactions and obtaining the public key value from the data sent. 

Reviewer 2: ```suggestion
This RKSIP proposes changing the Bridge code of RegisterBtcTransactions function. This function is responsible for parsing Bridge transactions and obtaining the public key value from the data sent. 
```

> RegisterBtcTransactions performs a number of validations to the BTC transaction to verify it legitimacy. After those validations, it will also check if it's a lock, and in case it is, it will then perform the lock using the sender's public key. These last two steps need to be changed.

Reviewer: `RegisterBtcTransactions` performs a number of validations to the BTC transaction to verify ~~it~~ **it's** legitimacy.

Reviewer: After those validations **have passed,** it ~~will~~ also check **(s)** if it's a lock, ~~and in case it is,~~ in that case it ~~will then~~ perform **(s)** the lock using the sender's public key. These last two steps **are the ones that** need to be changed.

> • Lock check. This check is only considering the input format for P2KH.

Reviewer: • Lock check. This check is **currently** only considering the input format for P2KH.

> • Sender public key fetching. Before getting the public key, this step will validate the input's format, only using the P2KH format.

Reviewer: Sender public key fetching. Before getting the public key, this step ~~will~~ validate **(s)** the input's format, **accepting** only ~~using~~ the P2KH format.

> In order to consider P2SH-PWPKH transactions, asuming the result of checking P2PK transcription legitmy as invalid, it should be checked if it is of P2SH-PWPKH type of transaction. The RegisterBtcTransactions function needs to verify lock check and gets the public key format using Segregated Witness format defined in BIP 144 [2].   

Reviewer: In order to consider P2SH-PWPKH transactions, asuming the result of checking P2PK transcription ~~legitmy~~ ** legitimacy** as invalid, it should be checked if it is ~~of~~ **a** P2SH-PWPKH type of transaction. The `RegisterBtcTransactions` function needs to verify lock check and get ~~s~~ the public key format using Segregated Witness format defined in BIP 144 [2].

Reviewer: **This is the only change required to be able to accept P2SH-PWPKH segwit transactions.**

> In order to consider P2SH-PWPKH transactions, assuming it is not possible to get the P2PKH public key from the input, it should be checked if it is a P2SH-PWPKH type of transaction. The RegisterBtcTransactions function needs to vverify if it is a lock transaction and get the public key format using Segregated Witness format defined in BIP 144 [2].   

Reviewer 2: ```suggestion
Assuming it is not possible to get the P2PKH public key from the transaction's input, it should be checked if it is a P2SH-PWPKH type of input. The RegisterBtcTransactions function needs to verify if it is a lock transaction and then get the public key format using Segregated Witness format defined in BIP 144 [2].   
```

> The only change required to be able to accept P2SH-PWPKH is what must be done in the function just mentioned.

Reviewer: ~~The only change required to be able to accept P2SH-PWPKH is what must be done in the function just mentioned.~~
